python/housinginsights/sources/base.py

```diff
@@ -164,7 +165,95 @@ def directly_to_file(self, data, filepath):
+        those fields in the given fields map.
+        :param nlihc_id: the nlihc_id for the current line in dchousing data
+        :param fields: the column headers data we want returned
+        :param fields_map: a dictionary mapping for the dchousing headers to
+        the field headers that should be returned
+        :param line: the current line in the dchousing data to be mapped
+        """
+        data = {}
+        for field in fields:
+            value = fields_map[field]
+            if value is None:
+                if field == 'Nlihc_id':
+                    data[field] = nlihc_id
+                else:
+                    data[field] = None
+            else:
+                data[field] = line[value]
+        return data
+
+    def _get_nlihc_id_from_db(self, db_conn, address_id):
+        """
+        Returns a tuple nlihc_id, in_proj_table_flag pair for a given address
+        id by performing a lookup on the project table in the database using
+        the mar_id column.
+
+        If the address id doesn't map to an existing building in the table,
+        a randomly generated uuid is returned as nlihc_id.
+        """
+        query = "select nlihc_id from project where mar_id = " \
+                "'{}';".format(address_id)
+        query_result = db_conn.execute(query)
+        result = [dict(x) for x in query_result.fetchall()]
+
+        if result:
+            return result[0]['nlihc_id'], True
+        else:
+            return str(uuid4()), False
 
+    def create_project_subsidy_csv(self, uid, project_fields_map,
+                                   subsidy_fields_map, database_choice=None):
+        """
+        Writes 'source_project' and 'source_subsidy' raw files from the
+        source csv file. It then deletes the source file so it
+        doesn't get added to manifest and loaded into the database.
+        """
+        if database_choice is None:
+            database_choice = 'docker_database'
+        engine = dbtools.get_database_engine(database_choice)
+        db_conn = engine.connect()
+
+        # create file path objects
+        source_csv = self.output_paths[uid]
+        folder = os.path.dirname(source_csv)
+        source_proj = os.path.join(folder, "{}_project.csv".format(uid))
+        source_subsidy = os.path.join(folder, "{}_subsidy.csv".format(uid))
+
+        # create dchousing_project/subsidy files from source
+        with open(source_csv, encoding='utf-8') as f, \
+                open(source_proj, mode='w', encoding='utf-8') as proj, \
+                open(source_subsidy, mode='w', encoding='utf-8') as subsidy:
+
+            reader = csv.DictReader(f)
+            proj_writer = csv.DictWriter(proj, fieldnames=PROJ_FIELDS)
+            proj_writer.writeheader()
+            subsidy_writer = csv.DictWriter(subsidy, fieldnames=SUBSIDY_FIELDS)
+            subsidy_writer.writeheader()
+
+            for line in reader:
```

Comment:
Add some comments to show reason - makes sense to me but we've discussed how these things interact a lot, would be confusing to a new person.

"If the project doesn't exists in the database, we want to add a new record to the project table"

"All projects need a new record added to the subsidy table. If that project already exists in the database, link it to the project using the nlihc_id; if not, link it to the record that was added to the proj_writer output file."